### PR TITLE
🐛 [amp-base-carousel 0.1] E2E tests for amp carousel arrows in lightbox

### LINKED ISSUE #33287: [amp-carousel]control buttons are out of sync with the slide numbers in the amp-carousel
## What's the issue?

When amp-carousel is used inside amp-lightbox, left and right arrow buttons are out of sync with the slide numbers in the carousel.

I expect the following behavior from amp-carousel, but it doesn't seem to work that way if I use amp-carousel inside amp-lightbox.
- when the first slide is displayed in amp-carousel, only the right button to be displayed
- when the last slide is displayed in amp-carousel, only the left button to be displayed
- In other cases, both left and right buttons will be displayed

## How do we reproduce the issue?

1. open [my playground](https://playground.amp.dev/#share=PCFET0NUWVBFIGh0bWw+CjxodG1sIOKaoT4KICA8aGVhZD4KICAgIDxtZXRhIGNoYXJzZXQ9InV0Zi04IiAvPgogICAgPHRpdGxlPmFtcC1saWdodGJveC1jYXJvdXNlbDwvdGl0bGU+CiAgICA8bGluayByZWw9ImNhbm9uaWNhbCIgaHJlZj0iYW1wcy5odG1sIiAvPgogICAgPG1ldGEKICAgICAgbmFtZT0idmlld3BvcnQiCiAgICAgIGNvbnRlbnQ9IndpZHRoPWRldmljZS13aWR0aCxtaW5pbXVtLXNjYWxlPTEsaW5pdGlhbC1zY2FsZT0xIgogICAgLz4KICAgIDxzY3JpcHQKICAgICAgYXN5bmMKICAgICAgY3VzdG9tLWVsZW1lbnQ9ImFtcC1saWdodGJveCIKICAgICAgc3JjPSJodHRwczovL2Nkbi5hbXBwcm9qZWN0Lm9yZy92MC9hbXAtbGlnaHRib3gtMC4xLmpzIgogICAgPjwvc2NyaXB0PgogICAgPHNjcmlwdAogICAgICBhc3luYwogICAgICBjdXN0b20tZWxlbWVudD0iYW1wLWNhcm91c2VsIgogICAgICBzcmM9Imh0dHBzOi8vY2RuLmFtcHByb2plY3Qub3JnL3YwL2FtcC1jYXJvdXNlbC0wLjIuanMiCiAgICA+PC9zY3JpcHQ+CiAgICA8c2NyaXB0CiAgICAgIGFzeW5jCiAgICAgIGN1c3RvbS1lbGVtZW50PSJhbXAtc2VsZWN0b3IiCiAgICAgIHNyYz0iaHR0cHM6Ly9jZG4uYW1wcHJvamVjdC5vcmcvdjAvYW1wLXNlbGVjdG9yLTAuMS5qcyIKICAgID48L3NjcmlwdD4KICAgIDxzdHlsZSBhbXAtY3VzdG9tPgogICAgICBib2R5IHsKICAgICAgICBwYWRkaW5nOiAxMHB4OwogICAgICB9CiAgICAgIC5pbm5lciB7CiAgICAgICAgZGlzcGxheTogZmxleDsKICAgICAgICBnYXA6IDEwcHg7CiAgICAgIH0KICAgICAgLmltYWdlIHsKICAgICAgICB3aWR0aDogMzAwcHg7CiAgICAgIH0KICAgICAgLmltYWdlICsgLmltYWdlIHsKICAgICAgICBtYXJnaW4tbGVmdDogMTBweDsKICAgICAgfQogICAgICAubGlnaHRib3ggewogICAgICAgIGJhY2tncm91bmQtY29sb3I6IGJsYWNrOwogICAgICB9CiAgICAgIC5jbG9zZS1idXR0b24gewogICAgICAgIHBvc2l0aW9uOiBhYnNvbHV0ZTsKICAgICAgICB6LWluZGV4OiAxOwogICAgICB9CiAgICA8L3N0eWxlPgogICAgPHN0eWxlIGFtcC1ib2lsZXJwbGF0ZT5ib2R5ey13ZWJraXQtYW5pbWF0aW9uOi1hbXAtc3RhcnQgOHMgc3RlcHMoMSxlbmQpIDBzIDEgbm9ybWFsIGJvdGg7LW1vei1hbmltYXRpb246LWFtcC1zdGFydCA4cyBzdGVwcygxLGVuZCkgMHMgMSBub3JtYWwgYm90aDstbXMtYW5pbWF0aW9uOi1hbXAtc3RhcnQgOHMgc3RlcHMoMSxlbmQpIDBzIDEgbm9ybWFsIGJvdGg7YW5pbWF0aW9uOi1hbXAtc3RhcnQgOHMgc3RlcHMoMSxlbmQpIDBzIDEgbm9ybWFsIGJvdGh9QC13ZWJraXQta2V5ZnJhbWVzIC1hbXAtc3RhcnR7ZnJvbXt2aXNpYmlsaXR5OmhpZGRlbn10b3t2aXNpYmlsaXR5OnZpc2libGV9fUAtbW96LWtleWZyYW1lcyAtYW1wLXN0YXJ0e2Zyb217dmlzaWJpbGl0eTpoaWRkZW59dG97dmlzaWJpbGl0eTp2aXNpYmxlfX1ALW1zLWtleWZyYW1lcyAtYW1wLXN0YXJ0e2Zyb217dmlzaWJpbGl0eTpoaWRkZW59dG97dmlzaWJpbGl0eTp2aXNpYmxlfX1ALW8ta2V5ZnJhbWVzIC1hbXAtc3RhcnR7ZnJvbXt2aXNpYmlsaXR5OmhpZGRlbn10b3t2aXNpYmlsaXR5OnZpc2libGV9fUBrZXlmcmFtZXMgLWFtcC1zdGFydHtmcm9te3Zpc2liaWxpdHk6aGlkZGVufXRve3Zpc2liaWxpdHk6dmlzaWJsZX19PC9zdHlsZT48bm9zY3JpcHQ+PHN0eWxlIGFtcC1ib2lsZXJwbGF0ZT5ib2R5ey13ZWJraXQtYW5pbWF0aW9uOm5vbmU7LW1vei1hbmltYXRpb246bm9uZTstbXMtYW5pbWF0aW9uOm5vbmU7YW5pbWF0aW9uOm5vbmV9PC9zdHlsZT48L25vc2NyaXB0PgogICAgPHNjcmlwdCBhc3luYyBzcmM9Imh0dHBzOi8vY2RuLmFtcHByb2plY3Qub3JnL3YwLmpzIj48L3NjcmlwdD4KICA8L2hlYWQ+CiAgPGJvZHk+CiAgICA8aDI+YW1wLWxpZ2h0Ym94LWNhcm91c2VsPC9oMj4KICAgIDxhbXAtc2VsZWN0b3IKICAgICAgaWQ9InNlbGVjdG9yMSIKICAgICAgb249InNlbGVjdDpjYXJvdXNlbDEuZ29Ub1NsaWRlKGluZGV4PWV2ZW50LnRhcmdldE9wdGlvbiksIGxpZ2h0Ym94MSIKICAgID4KICAgICAgPGRpdiBjbGFzcz0iaW5uZXIiPgogICAgICAgIDxhbXAtaW1nCiAgICAgICAgICBjbGFzcz0iaW1hZ2UiCiAgICAgICAgICBzcmM9Imh0dHA6Ly9wbGFjZWtpdHRlbi5jb20vNDUwLzMwMD9pbWFnZT0xIgogICAgICAgICAgb3B0aW9uPSIwIgogICAgICAgICAgbGF5b3V0PSJyZXNwb25zaXZlIgogICAgICAgICAgd2lkdGg9IjQ1MCIKICAgICAgICAgIGhlaWdodD0iMzAwIgogICAgICAgID48L2FtcC1pbWc+CiAgICAgICAgPGFtcC1pbWcKICAgICAgICAgIGNsYXNzPSJpbWFnZSIKICAgICAgICAgIHNyYz0iaHR0cDovL3BsYWNla2l0dGVuLmNvbS80NTAvMzAwP2ltYWdlPTIiCiAgICAgICAgICBsYXlvdXQ9InJlc3BvbnNpdmUiCiAgICAgICAgICBvcHRpb249IjEiCiAgICAgICAgICB3aWR0aD0iNDUwIgogICAgICAgICAgaGVpZ2h0PSIzMDAiCiAgICAgICAgPjwvYW1wLWltZz4KICAgICAgICA8YW1wLWltZwogICAgICAgICAgY2xhc3M9ImltYWdlIgogICAgICAgICAgc3JjPSJodHRwOi8vcGxhY2VraXR0ZW4uY29tLzQ1MC8zMDA/aW1hZ2U9MyIKICAgICAgICAgIGxheW91dD0icmVzcG9uc2l2ZSIKICAgICAgICAgIG9wdGlvbj0iMiIKICAgICAgICAgIHdpZHRoPSI0NTAiCiAgICAgICAgICBoZWlnaHQ9IjMwMCIKICAgICAgICA+PC9hbXAtaW1nPgogICAgICAgIDxhbXAtaW1nCiAgICAgICAgICBjbGFzcz0iaW1hZ2UiCiAgICAgICAgICBzcmM9Imh0dHA6Ly9wbGFjZWtpdHRlbi5jb20vNDUwLzMwMD9pbWFnZT00IgogICAgICAgICAgbGF5b3V0PSJyZXNwb25zaXZlIgogICAgICAgICAgb3B0aW9uPSIzIgogICAgICAgICAgd2lkdGg9IjQ1MCIKICAgICAgICAgIGhlaWdodD0iMzAwIgogICAgICAgID48L2FtcC1pbWc+CiAgICAgIDwvZGl2PgogICAgPC9hbXAtc2VsZWN0b3I+CiAgICA8YW1wLWxpZ2h0Ym94IGlkPSJsaWdodGJveDEiIGNsYXNzPSJsaWdodGJveCIgbGF5b3V0PSJub2Rpc3BsYXkiPgogICAgICA8YnV0dG9uIGNsYXNzPSJjbG9zZS1idXR0b24iIG9uPSJ0YXA6bGlnaHRib3gxLmNsb3NlIj5jbG9zZTwvYnV0dG9uPgogICAgICA8YW1wLWNhcm91c2VsCiAgICAgICAgaWQ9ImNhcm91c2VsMSIKICAgICAgICB3aWR0aD0iOTAwIgogICAgICAgIGhlaWdodD0iNjAwIgogICAgICAgIGxheW91dD0iZmlsbCIKICAgICAgICB0eXBlPSJzbGlkZXMiCiAgICAgICAgb249InNsaWRlQ2hhbmdlOnNlbGVjdG9yMS50b2dnbGUoaW5kZXg9ZXZlbnQuaW5kZXgsIHZhbHVlPXRydWUpIgogICAgICA+CiAgICAgICAgPGFtcC1pbWcKICAgICAgICAgIHNyYz0iaHR0cDovL3BsYWNla2l0dGVuLmNvbS85MDAvNjAwP2ltYWdlPTEiCiAgICAgICAgICBsYXlvdXQ9InJlc3BvbnNpdmUiCiAgICAgICAgICB3aWR0aD0iOTAwIgogICAgICAgICAgaGVpZ2h0PSI2MDAiCiAgICAgICAgPjwvYW1wLWltZz4KICAgICAgICA8YW1wLWltZwogICAgICAgICAgc3JjPSJodHRwOi8vcGxhY2VraXR0ZW4uY29tLzkwMC82MDA/aW1hZ2U9MiIKICAgICAgICAgIGxheW91dD0icmVzcG9uc2l2ZSIKICAgICAgICAgIHdpZHRoPSI5MDAiCiAgICAgICAgICBoZWlnaHQ9IjYwMCIKICAgICAgICA+PC9hbXAtaW1nPgogICAgICAgIDxhbXAtaW1nCiAgICAgICAgICBzcmM9Imh0dHA6Ly9wbGFjZWtpdHRlbi5jb20vOTAwLzYwMD9pbWFnZT0zIgogICAgICAgICAgbGF5b3V0PSJyZXNwb25zaXZlIgogICAgICAgICAgd2lkdGg9IjkwMCIKICAgICAgICAgIGhlaWdodD0iNjAwIgogICAgICAgID48L2FtcC1pbWc+CiAgICAgICAgPGFtcC1pbWcKICAgICAgICAgIHNyYz0iaHR0cDovL3BsYWNla2l0dGVuLmNvbS85MDAvNjAwP2ltYWdlPTQiCiAgICAgICAgICBsYXlvdXQ9InJlc3BvbnNpdmUiCiAgICAgICAgICB3aWR0aD0iOTAwIgogICAgICAgICAgaGVpZ2h0PSI2MDAiCiAgICAgICAgPjwvYW1wLWltZz4KICAgICAgPC9hbXAtY2Fyb3VzZWw+CiAgICA8L2FtcC1saWdodGJveD4KICA8L2JvZHk+CjwvaHRtbD4K).
2. click the second image.
スクリーンショット 2021-03-16 13 45 08

3. the lightbox will open and the second image will be displayed, but you can see that only the right  button is displayed, the left button is not.

## What browsers are affected?
All browsers.

## Which AMP version is affected?
- LTS version of amp
- amp-carousel v0.2
- amp-lightbox v0.1
- amp-selector v0.1

### PR DESCRIPTION
**Pull request by micajuine-ho**
Seems like the original issue was fixed via some other PR (probably runtime). This PR will now be to add e2e tests for this behavior.
~Closes #33287~

### CODE REVIEW

File: extensions/amp-base-carousel/0.1/carousel.js [hunk 1]
```diff
@@ -940,6 +940,10 @@ export class Carousel {
    * @private
    */
   isScrollAtEndingEdge_() {
+    // Lightbox and trivial case
+    if (!this.mixedLength_ && this.visibleCount_ === 1) {
+      return this.getCurrentIndex() >= this.slides_.length - 1;
+    }
```
Comment (caroqliu) [inline extensions/amp-base-carousel/0.1/carousel.js:946]: Shouldn't this check go in `isAtEnd` instead of `isScrollAtEndingEdge_`, since that is where the `loop` guard is also in place?

File: extensions/amp-base-carousel/0.1/carousel.js [hunk 2]
```diff
@@ -961,6 +965,10 @@ export class Carousel {
    * @private
    */
   isScrollAtBeginningEdge_() {
+    // Lightbox and trivial case
+    if (!this.mixedLength_ && this.visibleCount_ === 1) {
+      return this.getCurrentIndex() <= 0;
+    }
```
Comment (caroqliu) [inline extensions/amp-base-carousel/0.1/carousel.js:971]: Same Q as above but with `isAtStart`

File: extensions/amp-base-carousel/0.1/test/test-carousel.js [hunk 3]
```diff
@@ -198,7 +204,9 @@ describes.realWin('carousel implementation', {}, (env) => {
         slideCount: 3,
         loop: false,
       });
-      carousel.goToSlide(1, {smoothScroll: false});
+
+      // Fake the scroll
```
Comment (caroqliu) [inline extensions/amp-base-carousel/0.1/test/test-carousel.js:208]: Why is this change necessary? If possible we should avoid overwriting internal state in tests.

File: extensions/amp-base-carousel/0.1/test/test-carousel.js [hunk 4]
```diff
@@ -298,23 +308,23 @@ describes.realWin('carousel implementation', {}, (env) => {
         slideCount: 3,
         initialIndex: -1,
       });
-      expect(carousel.isAtStart()).to.be.true;
+      expect(carousel.scrollContainer_.scrollLeft).to.equal(0);
```
Comment (caroqliu) [inline extensions/amp-base-carousel/0.1/test/test-carousel.js:311]: Is this still needed if we modify `isAtStart()` instead?

File: extensions/amp-carousel/0.2/test/test-e2e/test-arrows-in-lightbox.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    beforeEach(async () => {
+      controller = env.controller;
+    });
+
+    it('should open with both arrows', async () => {
+      // Click on image 2
+      const secondImage = await controller.findElement('#second');
+      await controller.click(secondImage);
+      // Wait for lightbox to load the carousel and image
+      await sleep(500);
```
Comment (caroqliu) [inline extensions/amp-carousel/0.2/test/test-e2e/test-arrows-in-lightbox.js:51]: Is there some other way to know precisely when the lightbox has loaded?

File: extensions/amp-carousel/0.2/test/test-e2e/test-arrows-in-lightbox.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    });
+
+    it('should open with one arrow', async () => {
+      // Click on image 2
+      const secondImage = await controller.findElement('#fourth');
+      await controller.click(secondImage);
+      // Wait for lightbox to load the carousel and image
+      await sleep(500);
+
+      // Both arrows should be showing
```
Comment (caroqliu) [inline extensions/amp-carousel/0.2/test/test-e2e/test-arrows-in-lightbox.js:71]:
```suggestion
      // Only prev arrow should be showing since non-looping carousel is on the last slide.
```

File: extensions/amp-carousel/0.2/test/test-e2e/test-arrows-in-lightbox.js [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {sleep} from './helpers';
+
+describes.endtoend(
+  'amp carousel in lightbox',
```
Comment (caroqliu) [inline extensions/amp-carousel/0.2/test/test-e2e/test-arrows-in-lightbox.js:20]:
```suggestion
  'amp carousel:0.2 in lightbox',
```

Comment (caroqliu) [reply extensions/amp-carousel/0.2/test/test-e2e/test-arrows-in-lightbox.js:20]: WDYT about adding an equivalent test suite for amp-base-carousel?

**Review by caroqliu — CHANGES_REQUESTED** [review]

File: extensions/amp-base-carousel/0.1/carousel.js (hunk 1 above)
Comment (micajuine-ho) [reply extensions/amp-base-carousel/0.1/carousel.js:946]: Great idea!

File: extensions/amp-base-carousel/0.1/test/test-carousel.js (hunk 3 above)
Comment (micajuine-ho) [reply extensions/amp-base-carousel/0.1/test/test-carousel.js:208]: This change is necessary since we changed `isAtStart` & `isAtEnd` to also use current index in it's calculation and that only get updates once the scroll event occurs (via listener), which is hard to test. 

I'll change the tests to also ensure that the scroll position is also correct (since that's what `isAtStart` and `isAtEnd` were essentially testing before).

File: extensions/amp-base-carousel/0.1/test/test-carousel.js (hunk 4 above)
Comment (micajuine-ho) [reply extensions/amp-base-carousel/0.1/test/test-carousel.js:311]: I made this change for the same reason above, since I wanted to keep the test inline with what `isAtStart` testing. But I can add both in for completenesses sake.

File: extensions/amp-base-carousel/0.1/test/test-carousel.js [hunk 8]
```diff
@@ -298,23 +308,23 @@ describes.realWin('carousel implementation', {}, (env) => {
...
     });
 
     it('should start at slide 0 with initialIndex that is greater than number of slides', async () => {
       const carousel = await createCarousel({
         slideCount: 3,
         initialIndex: 4,
       });
-      expect(carousel.isAtStart()).to.be.true;
+      expect(carousel.scrollContainer_.scrollLeft).to.equal(0);
```
Comment (micajuine-ho) [inline extensions/amp-base-carousel/0.1/test/test-carousel.js:319]: @caroqliu So actually, using initial index 4 breaks amp-base-carousel: [repro](https://playground.amp.dev/?mode=Responsive#share=PCFET0NUWVBFIGh0bWw+CjxodG1sIOKaoT4KICA8aGVhZD4KICAgIDxtZXRhIGNoYXJzZXQ9InV0Zi04IiAvPgogICAgPHRpdGxlPkFNUCBMaWdodGJveCBDYXJvdXNlbCBXaXRoIEFNUCBTZWxlY3RvcjwvdGl0bGU+CiAgICA8bGluayByZWw9ImNhbm9uaWNhbCIgaHJlZj0ic2VsZi5odG1sIiAvPgogICAgPG1ldGEgbmFtZT0idmlld3BvcnQiIGNvbnRlbnQ9IndpZHRoPWRldmljZS13aWR0aCIgLz4KICAgIDxzdHlsZSBhbXAtYm9pbGVycGxhdGU+Ym9keXstd2Via2l0LWFuaW1hdGlvbjotYW1wLXN0YXJ0IDhzIHN0ZXBzKDEsZW5kKSAwcyAxIG5vcm1hbCBib3RoOy1tb3otYW5pbWF0aW9uOi1hbXAtc3RhcnQgOHMgc3RlcHMoMSxlbmQpIDBzIDEgbm9ybWFsIGJvdGg7LW1zLWFuaW1hdGlvbjotYW1wLXN0YXJ0IDhzIHN0ZXBzKDEsZW5kKSAwcyAxIG5vcm1hbCBib3RoO2FuaW1hdGlvbjotYW1wLXN0YXJ0IDhzIHN0ZXBzKDEsZW5kKSAwcyAxIG5vcm1hbCBib3RofUAtd2Via2l0LWtleWZyYW1lcyAtYW1wLXN0YXJ0e2Zyb217dmlzaWJpbGl0eTpoaWRkZW59dG97dmlzaWJpbGl0eTp2aXNpYmxlfX1ALW1vei1rZXlmcmFtZXMgLWFtcC1zdGFydHtmcm9te3Zpc2liaWxpdHk6aGlkZGVufXRve3Zpc2liaWxpdHk6dmlzaWJsZX19QC1tcy1rZXlmcmFtZXMgLWFtcC1zdGFydHtmcm9te3Zpc2liaWxpdHk6aGlkZGVufXRve3Zpc2liaWxpdHk6dmlzaWJsZX19QC1vLWtleWZyYW1lcyAtYW1wLXN0YXJ0e2Zyb217dmlzaWJpbGl0eTpoaWRkZW59dG97dmlzaWJpbGl0eTp2aXNpYmxlfX1Aa2V5ZnJhbWVzIC1hbXAtc3RhcnR7ZnJvbXt2aXNpYmlsaXR5OmhpZGRlbn10b3t2aXNpYmlsaXR5OnZpc2libGV9fTwvc3R5bGU+PG5vc2NyaXB0PjxzdHlsZSBhbXAtYm9pbGVycGxhdGU+Ym9keXstd2Via2l0LWFuaW1hdGlvbjpub25lOy1tb3otYW5pbWF0aW9uOm5vbmU7LW1zLWFuaW1hdGlvbjpub25lO2FuaW1hdGlvbjpub25lfTwvc3R5bGU+PC9ub3NjcmlwdD4KICAgIDxzY3JpcHQgYXN5bmMgc3JjPSJodHRwczovL2Nkbi5hbXBwcm9qZWN0Lm9yZy92MC5qcyI+PC9zY3JpcHQ+CiAgICA8c2NyaXB0IGFzeW5jIGN1c3RvbS1lbGVtZW50PSJhbXAtY2Fyb3VzZWwiIHNyYz0iaHR0cHM6Ly9jZG4uYW1wcHJvamVjdC5vcmcvdjAvYW1wLWNhcm91c2VsLTAuMi5qcyI+PC9zY3JpcHQ+CiAgICA8c3R5bGUgYW1wLWN1c3RvbT4KICAgIAogICAgPC9zdHlsZT4KICA8L2hlYWQ+CiAgPGJvZHk+CiAgPGFtcC1jYXJvdXNlbCBpZD0iY2Fyb3VzZWwxIiB3aWR0aD0iNDUwIiBoZWlnaHQ9IjMwMCIgdHlwZT0ic2xpZGVzIiBzbGlkZT0iNCI+CiAgICAgICAgICA8YW1wLWltZyBzcmM9Ii9zdGF0aWMvaW5saW5lLWV4YW1wbGVzL2ltYWdlcy9pbWFnZTEuanBnIgogICAgd2lkdGg9IjQ1MCIKICAgIGhlaWdodD0iMzAwIj48L2FtcC1pbWc+CiAgPGFtcC1pbWcgc3JjPSIvc3RhdGljL2lubGluZS1leGFtcGxlcy9pbWFnZXMvaW1hZ2UyLmpwZyIKICAgIHdpZHRoPSI0NTAiCiAgICBoZWlnaHQ9IjMwMCI+PC9hbXAtaW1nPgogIDxhbXAtaW1nIHNyYz0iL3N0YXRpYy9pbmxpbmUtZXhhbXBsZXMvaW1hZ2VzL2ltYWdlMy5qcGciCiAgICB3aWR0aD0iNDUwIgogICAgaGVpZ2h0PSIzMDAiPjwvYW1wLWltZz4KICA8L2FtcC1jYXJvdXNlbD4KPC9ib2R5Pgo8L2h0bWw+Cg==)

I've filed https://github.com/ampproject/amphtml/issues/33512 to tackle this problem in amp-carousel 0.2. 

IMO we either remove this test or fix it in this PR or fix it in a later PR.

File: extensions/amp-carousel/0.2/test/test-e2e/test-arrows-in-lightbox.js (hunk 5 above)
Comment (micajuine-ho) [reply extensions/amp-carousel/0.2/test/test-e2e/test-arrows-in-lightbox.js]: Yes, found it: `await expect(await controller.getElementProperty(lightbox, 'style')).to.not.be.null;`

Note that 
1.) Using hidden is not good here (checking for hidden to be an element property), since it's only added *after* it's closed,  so not good to check initially that it's opened
2.) You must you `await` for `controller.getElementProperty`, as with all controller methods, it returns a promise
- I found a few e2e tests that don't wait for this which probably means they're not doing what they should be.

File: extensions/amp-carousel/0.2/test/test-e2e/test-arrows-in-lightbox.js (hunk 7 above)
Comment (micajuine-ho) [reply extensions/amp-carousel/0.2/test/test-e2e/test-arrows-in-lightbox.js:18]: Is it recommended that devs use amp-base-carousel in a Lightbox?

File: extensions/amp-base-carousel/0.1/test/test-carousel.js (hunk 3 above)
Comment (caroqliu) [reply extensions/amp-base-carousel/0.1/test/test-carousel.js]: Can we send a fake scroll event instead of overwriting internal state?

File: extensions/amp-carousel/0.2/test/test-e2e/test-arrows-in-lightbox.js (hunk 7 above)
Comment (caroqliu) [reply extensions/amp-carousel/0.2/test/test-e2e/test-arrows-in-lightbox.js:18]: Yes both are possible.

File: extensions/amp-base-carousel/0.1/test/test-carousel.js (hunk 8 above)
Comment (caroqliu) [reply extensions/amp-base-carousel/0.1/test/test-carousel.js]: Wanted to unblock you here, fixed in #33512.

**Review by caroqliu — APPROVED** [review]
